diff --git a/hw/riscv/boot.c b/hw/riscv/boot.c
index 75f34287ff..44562e2646 100644
--- a/hw/riscv/boot.c
+++ b/hw/riscv/boot.c
@@ -26,6 +26,7 @@
 #include "hw/loader.h"
 #include "hw/riscv/boot.h"
 #include "hw/riscv/boot_opensbi.h"
+#include "hw/riscv/virt.h"
 #include "elf.h"
 #include "system/device_tree.h"
 #include "system/qtest.h"
@@ -438,6 +439,39 @@ void riscv_setup_rom_reset_vec(MachineState *machine, RISCVHartArrayState *harts
     int i;
     uint32_t start_addr_hi32 = 0x00000000;
     uint32_t fdt_load_addr_hi32 = 0x00000000;
+    RISCVVirtState *rvs = RISCV_VIRT_MACHINE(machine);
+
+    if (rvs->rom) {
+        uint8_t *rom = g_malloc(rom_size);
+        if (!rom) {
+            error_report("could not allocate rom buffer");
+            exit(1);
+        }
+
+        size_t bl_size = load_image_size(rvs->rom, rom, rom_size);
+        if (bl_size == -1) {
+            error_report("could not get rom size");
+            exit(1);
+        }
+
+        rom_add_blob_fixed_as("mrom.reset", rom, bl_size,
+                                rom_base, &address_space_memory);
+        g_free(rom);
+
+        /* we also need the fdt */
+        if (fdt_pack(machine->fdt) ||
+                fdt_totalsize(machine->fdt) > rom_size - bl_size) {
+            error_report("not enough space to store device tree");
+            exit(1);
+        }
+
+        // word-aligned fdt location
+        size_t fdt_loc = rom_base + bl_size;
+        fdt_loc = ((fdt_loc + 0x4 - 1) & ~(0x4 - 1));
+        rom_add_blob_fixed_as("mrom.fdt", machine->fdt, fdt_totalsize(machine->fdt),
+                                fdt_loc, &address_space_memory);
+        return;
+    }
 
     if (!riscv_is_32bit(harts)) {
         start_addr_hi32 = start_addr >> 32;
diff --git a/hw/riscv/virt.c b/hw/riscv/virt.c
index 17909206c7..b779f61bbc 100644
--- a/hw/riscv/virt.c
+++ b/hw/riscv/virt.c
@@ -1872,6 +1872,20 @@ static void virt_set_acpi(Object *obj, Visitor *v, const char *name,
     visit_type_OnOffAuto(v, name, &s->acpi, errp);
 }
 
+static char *virt_get_rom(Object *obj, Error **errp)
+{
+    RISCVVirtState *rvs = RISCV_VIRT_MACHINE(obj);
+    return g_strdup(rvs->rom);
+}
+
+static void virt_set_rom(Object *obj, const char *value, Error **errp)
+{
+    RISCVVirtState *rvs = RISCV_VIRT_MACHINE(obj);
+
+    g_free(rvs->rom);
+    rvs->rom = g_strdup(value);
+}
+
 static HotplugHandler *virt_machine_get_hotplug_handler(MachineState *machine,
                                                         DeviceState *dev)
 {
@@ -1959,6 +1973,10 @@ static void virt_machine_class_init(ObjectClass *oc, const void *data)
     object_class_property_add_str(oc, "aia-guests",
                                   virt_get_aia_guests,
                                   virt_set_aia_guests);
+
+    object_class_property_add_str(oc, "rom", virt_get_rom, virt_set_rom);
+    object_class_property_set_description(oc, "rom", "ROM image");
+
     {
         g_autofree char *str =
             g_strdup_printf("Set number of guest MMIO pages for AIA IMSIC. "
diff --git a/include/hw/riscv/virt.h b/include/hw/riscv/virt.h
index 7b4c2c8b7d..ff59606805 100644
--- a/include/hw/riscv/virt.h
+++ b/include/hw/riscv/virt.h
@@ -64,6 +64,7 @@ struct RISCVVirtState {
     struct GPEXHost *gpex_host;
     OnOffAuto iommu_sys;
     uint16_t pci_iommu_bdf;
+    char *rom;
 };
 
 enum {
